#include "pay_window.h"
#include "ui_pay_window.h"
#include "boarderitem.h"
#include "interfacemanager.h"
#include "flightinfo.h"
#include "listitem.h"
#include "usermanager.h"
#include <QListWidgetItem>
#include <QListWidget>
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QJsonDocument>
#include <QMessageBox>
#include <QLineEdit>

pay_window::pay_window(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::pay_window)
{
    ui->setupUi(this);
    QPixmap pixmap("://bcgrd.jpg");
    QPixmap scaledPixmap = pixmap.scaled(ui->label_5->size(),
                                         Qt::KeepAspectRatio,
                                         Qt::SmoothTransformation);
    ui->label_5->setPixmap(scaledPixmap);
    ui->btn_f1->setVisible(false);
    ui->btn_f2->setVisible(false);
    ui->btn_f3->setVisible(false);
}

pay_window::~pay_window()
{
    delete ui;
}

void pay_window::on_pushButton_clicked()
{
    boarderItem* li = new boarderItem();
    QListWidgetItem *item = new QListWidgetItem(ui->listWidget); // å‡è®¾ä½ çš„åˆ—è¡¨æ§ä»¶å« listWidget
    item->setSizeHint(li->sizeHint());
    ui->listWidget->addItem(item);
    ui->listWidget->setItemWidget(item, li);
    emit countChanged();
}

void pay_window:: loadFliInfo()
{
    ui->listWidget->clear();
    boarderItem::num=1;
    on_pushButton_clicked();
    flightInfo fli= InterfaceManager::instance()->Info;
    ui->lbl_compName->setText(fli.airlineCompany);
    const QPixmap airlogo=getPic(fli.airlineCompany);
    ui->lbl_compPic->setPixmap(airlogo);
    ui->lbl_depar->setText(fli.departure);
    ui->lbl_des->setText(fli.destination);
    ui->lbl_flightCate->setText(fli.aircraftModel);
    ui->lbl_flightNo->setText(fli.flightNumber);
    QDateTime departureDateTime = QDateTime::fromSecsSinceEpoch(fli.departureTime);
    QDateTime arrivalDateTime = QDateTime::fromSecsSinceEpoch(fli.arrivalTime);
    ui->lbl_deparTime->setText(departureDateTime.toString("hh:mm"));
    ui->lbl_arrivalTime->setText(arrivalDateTime.toString("hh:mm"));
    QLocale chineseLocale(QLocale::Chinese, QLocale::China);
    QString dayofweek=chineseLocale.toString(departureDateTime, "dddd");
    ui->lbl_date->setText(departureDateTime.toString("MM-dd")+" "+dayofweek);
    ui->lbl_deparAirport->setText(fli.departureAirport);
    ui->lbl_arrivalAirport->setText(fli.arrivalAirport);
    if(fli.chosenSeat=="ç»æµèˆ±") price=fli.economyClassPrice*ui->listWidget->count();
    else if(fli.chosenSeat=="å¤´ç­‰èˆ±") price=fli.firstClassPrice*ui->listWidget->count();
    ui->lbl_price->setText("Â¥" + QString::number(static_cast<int>(price)));
    ui->lbl_seatCate->setText(fli.chosenSeat);
    connect(this,&pay_window::countChanged,this,[this,fli](){
        if(fli.chosenSeat=="ç»æµèˆ±") price+=fli.economyClassPrice;
        else if(fli.chosenSeat=="å¤´ç­‰èˆ±") price+=fli.firstClassPrice;
        ui->lbl_price->setText("Â¥" + QString::number(static_cast<int>(price)));
    });
    connect(this,&pay_window::itemCountChanged,this,[this,fli](){
        ui->lbl_price->setText("Â¥" + QString::number(static_cast<int>(price)));
    });
}

const QPixmap pay_window:: getPic(const QString &name)
{
    QPixmap airlineLogo;
    if(name=="å››å·èˆªç©º") airlineLogo.load("://sc.png");
    if(name=="å¦é—¨èˆªç©º") airlineLogo.load("://xm.png");
    if(name=="æ·±åœ³èˆªç©º") airlineLogo.load("://sh.png");
    if(name=="å±±ä¸œèˆªç©º") airlineLogo.load("://sdh.png");
    if(name=="å—æ–¹èˆªç©º") airlineLogo.load("://nh.jpg");
    if(name=="æ˜†æ˜èˆªç©º") airlineLogo.load("://km.png");
    if(name=="ä¸­å›½è”åˆèˆªç©º") airlineLogo.load("://lh.png");
    if(name=="ä¸­å›½å›½èˆª") airlineLogo.load("://gh.png");
    if(name=="æµ·å—èˆªç©º") airlineLogo.load("://hnhk.png");
    if(name=="ä¸œæ–¹èˆªç©º") airlineLogo.load("://df.png");
    if(name=="å‰ç¥¥èˆªç©º") airlineLogo.load("://jx.png");
    if(name=="åå¤èˆªç©º") airlineLogo.load("://hxh.png");
    if(name=="é•¿é¾™èˆªç©º") airlineLogo.load("://cl.png");
    if(name=="é‡åº†èˆªç©º") airlineLogo.load("://cq.png");
    if(name=="è¥¿è—èˆªç©º") airlineLogo.load("://xz.png");
    if(name=="æˆéƒ½èˆªç©º") airlineLogo.load("://cd.png");
    if(name=="æµ·èˆªæ——ä¸‹è¥¿éƒ¨èˆª") airlineLogo.load("://hnhk.png");
    if(name=="æµ·èˆªæ——ä¸‹å¤©æ´¥èˆª") airlineLogo.load("://tj.png");
    if(name=="é¾™æ±Ÿèˆªç©º") airlineLogo.load("://lj.png");
    if(name=="è‹å—ç‘ä¸½èˆªç©º") airlineLogo.load("://rl.png");
    if(name=="æ²³åŒ—èˆªç©º") airlineLogo.load("://hb.png");
    if(name=="æµ·èˆªæ——ä¸‹åŒ—éƒ¨æ¹¾") airlineLogo.load("://bbw.png");
    if(name=="æµ·èˆªæ——ä¸‹ç¦èˆª") airlineLogo.load("://fh.png");
    if(name=="æµ·èˆªæ——ä¸‹ä¹Œèˆª") airlineLogo.load("://wh.png");
    if(name=="æµ·èˆªæ——ä¸‹é‡‘é¹èˆªç©º") airlineLogo.load("://jp.png");
    if(name=="æµ·èˆªæ——ä¸‹ç¥¥é¹èˆªç©º") airlineLogo.load("://xp.png");
    if(name=="é’å²›èˆªç©º") airlineLogo.load("://qd.png");
    if(name=="ä¸Šæµ·èˆªç©º") airlineLogo.load("://sh.png");
    if(name=="æµ·èˆªæ——ä¸‹é¦–éƒ½èˆªç©º") airlineLogo.load("://sd.png");
    if(name=="æµ·èˆªæ——ä¸‹é•¿å®‰èˆªç©º") airlineLogo.load("://cahk.png");
    if(name=="æ¹–å—èˆªç©º") airlineLogo.load("://hunan.png");
    airlineLogo = airlineLogo.scaled(QSize(51, 51), Qt::KeepAspectRatio);
    return airlineLogo;
}


void pay_window::on_pushButton_2_clicked()
{
        // éå†æ‰€æœ‰ä¹˜æœºäººä¿¡æ¯
        for(int i = 0; i < ui->listWidget->count(); i++) {
            QWidget* widget = ui->listWidget->itemWidget(ui->listWidget->item(i));
            if(widget) {
                // é€šè¿‡widgetæŸ¥æ‰¾å­æ§ä»¶
                QLineEdit* nameEdit = widget->findChild<QLineEdit*>("lineEdit");
                QLineEdit* idEdit = widget->findChild<QLineEdit*>("lineEdit_2");
                QLineEdit* phoneEdit = widget->findChild<QLineEdit*>("lineEdit_3");

                // æ£€æŸ¥å§“å
                if(nameEdit && nameEdit->text().isEmpty()) {
                    QMessageBox::warning(this, "æç¤º", "ä¹˜æœºäººå§“åä¸èƒ½ä¸ºç©ºï¼");
                    return;
                }
                // æ£€æŸ¥è¯ä»¶å·
                if(idEdit && idEdit->text().isEmpty()) {
                    QMessageBox::warning(this, "æç¤º", "ä¹˜æœºäººè¯ä»¶å·ä¸èƒ½ä¸ºç©ºï¼");
                    return;
                }
                // æ£€æŸ¥æ‰‹æœºå·
                if(phoneEdit && phoneEdit->text().isEmpty()) {
                    QMessageBox::warning(this, "æç¤º", "ä¹˜æœºäººæ‰‹æœºå·ä¸èƒ½ä¸ºç©ºï¼");
                    return;
                }
            }
        }

        // æ‰€æœ‰ä¿¡æ¯éªŒè¯é€šè¿‡åï¼Œåˆ‡æ¢é¡µé¢
        ui->stackedWidget->setCurrentIndex(1);
        ui->stackedWidget_2->setCurrentIndex(1);
}


void pay_window::on_btn_1_clicked()
{
    ui->btn_1->setVisible(false);
    ui->btn_f1->setVisible(true);
    price+=40*ui->listWidget->count();
    emit itemCountChanged();
}


void pay_window::on_btn_2_clicked()
{
    ui->btn_2->setVisible(false);
    ui->btn_f2->setVisible(true);
    price+=39*ui->listWidget->count();
    emit itemCountChanged();
}


void pay_window::on_btn_3_clicked()
{
    ui->btn_3->setVisible(false);
    ui->btn_f3->setVisible(true);
    price+=75*ui->listWidget->count();
    emit itemCountChanged();
}


void pay_window::on_btn_f1_clicked()
{
    ui->btn_f1->setVisible(false);
    ui->btn_1->setVisible(true);
    price-=40*ui->listWidget->count();
    emit itemCountChanged();
}


void pay_window::on_btn_f2_clicked()
{
    ui->btn_f2->setVisible(false);
    ui->btn_2->setVisible(true);
    price-=39*ui->listWidget->count();
    emit itemCountChanged();
}


void pay_window::on_btn_f3_clicked()
{
    ui->btn_f3->setVisible(false);
    ui->btn_3->setVisible(true);
    price-=75*ui->listWidget->count();
    emit itemCountChanged();
}


void pay_window::on_pushButton_8_clicked()
{
    ui->stackedWidget->setCurrentIndex(2);
    ui->stackedWidget_2->setCurrentIndex(2);
}


void pay_window::on_pushButton_3_clicked()
{
    QNetworkAccessManager* networkmanager = new QNetworkAccessManager();
    QNetworkRequest request(QUrl("http://127.0.0.1:8080/api/orders/create"));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");

    QJsonObject json;
    User currentUser = UserManager::getInstance()->getCurrentUser();
    json["userId"] = UserManager::getInstance()->getCurrentUser().id;
    json["flightId"] = InterfaceManager::instance()->Info.id;
    json["price"] = price; // ç¡®ä¿ä¸ºæµ®ç‚¹æ•°
    json["seatType"] =InterfaceManager::instance()->Info.chosenSeat; // æ”¹ä¸ºä¸­æ–‡

    QJsonDocument doc(json);
    qDebug() << "Request Body:" << doc.toJson();

    QNetworkReply* reply = networkmanager->post(request, doc.toJson());
    connect(reply, &QNetworkReply::finished, [this, reply,currentUser]() {
        QByteArray responseData = reply->readAll();
        qDebug() << "Response:" << responseData;

        if (reply->error() == QNetworkReply::NoError) {
            QJsonDocument jsonResponse = QJsonDocument::fromJson(responseData);
            QJsonObject jsonObject = jsonResponse.object();
            if (jsonObject["code"].toInt() == 200) {
                QMessageBox msgBox(this);
                msgBox.setWindowTitle("Success");
                msgBox.setText("è®¢å•æ·»åŠ æˆåŠŸ !");
                        QString email = currentUser.email;
                        QString subject = "âœˆï¸ æœºç¥¨é¢„è®¢æˆåŠŸé€šçŸ¥ ğŸ‰";

                        QString message = QString(
                                              "äº²çˆ±çš„ %1ï¼šğŸ¤—\n\n"
                                              "æ‚¨çš„æœºç¥¨é¢„è®¢å·²æˆåŠŸå®Œæˆï¼âœ¨ ä»¥ä¸‹æ˜¯æ‚¨çš„èˆªç­è¯¦ç»†ä¿¡æ¯ï¼š\n"
                                              "â”â”â”â”â”â”â”â”â”â” èˆªç­ä¿¡æ¯ â”â”â”â”â”â”â”â”â”â”\n\n"
                                              "ğŸ›« èˆªç­å·ï¼š%2\n"
                                              "ğŸ¢ èˆªç©ºå…¬å¸ï¼š%3\n"
                                              "ğŸ“ å‡ºå‘åœ°ï¼š%4 (%5)\n"
                                              "ğŸ¯ ç›®çš„åœ°ï¼š%6 (%7)\n"
                                              "â° å‡ºå‘æ—¶é—´ï¼š%8\n"
                                              "ğŸ•’ åˆ°è¾¾æ—¶é—´ï¼š%9\n"
                                              "ğŸ’º åº§ä½ç±»å‹ï¼š%10\n"
                                              "ğŸšª ç™»æœºå£ï¼š%11\n"
                                              "âœˆï¸ æœºå‹ï¼š%12\n\n"
                                              "â”â”â”â”â”â”â”â”â”â” é™„åŠ ä¿¡æ¯ â”â”â”â”â”â”â”â”â”â”\n\n"
                                              "ğŸ§³ è¡Œæé™åˆ¶ï¼š%13 kg\n"
                                              "ğŸ½ï¸ é¤é£ŸæœåŠ¡ï¼š%14\n"
                                              "ğŸŒ èˆªç­ç±»å‹ï¼š%15\n\n"
                                              "ğŸ’¡ æ¸©é¦¨æç¤ºï¼š\n"
                                              "â€¢ è¯·æå‰2å°æ—¶åˆ°è¾¾æœºåœºåŠç†å€¼æœºæ‰‹ç»­ â°\n"
                                              "â€¢ è¯·éšèº«æºå¸¦æœ‰æ•ˆèº«ä»½è¯ä»¶ ğŸ“„\n"
                                              "â€¢ å¦‚éœ€æ›´æ”¹è¡Œç¨‹è¯·æå‰è”ç³»å®¢æœ ğŸ“\n\n"
                                              "ç¥æ‚¨æ—…é€”æ„‰å¿«ï¼ğŸŒˆâœ¨\n\n"
                                              "æ­¤è‡´\n"
                                              "æœºç¥¨é¢„è®¢ç³»ç»Ÿ ğŸŠ\n\n"
                                              "â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\n"
                                              "æ³¨ï¼šæ­¤é‚®ä»¶ä¸ºç³»ç»Ÿè‡ªåŠ¨å‘é€ï¼Œè¯·å‹¿å›å¤ ğŸ“©"
                                              ).arg(currentUser.username)
                                              .arg(InterfaceManager::instance()->Info.flightNumber)
                                              .arg(InterfaceManager::instance()->Info.airlineCompany)
                                              .arg(InterfaceManager::instance()->Info.departure)
                                              .arg(InterfaceManager::instance()->Info.departureAirport)
                                              .arg(InterfaceManager::instance()->Info.destination)
                                              .arg(InterfaceManager::instance()->Info.arrivalAirport)
                                              .arg(QDateTime::fromSecsSinceEpoch(InterfaceManager::instance()->Info.departureTime).toString("yyyy-MM-dd hh:mm"))
                                              .arg(QDateTime::fromSecsSinceEpoch(InterfaceManager::instance()->Info.arrivalTime).toString("yyyy-MM-dd hh:mm"))
                                              .arg(InterfaceManager::instance()->Info.chosenSeat)
                                              .arg(InterfaceManager::instance()->Info.boardingGate)
                                              .arg(InterfaceManager::instance()->Info.aircraftModel)
                                              .arg(InterfaceManager::instance()->Info.luggageSizeLimit)
                                              .arg(InterfaceManager::instance()->Info.hasMeal ? "æä¾› ğŸ´" : "ä¸æä¾› âŒ")
                                              .arg(InterfaceManager::instance()->Info.isInternational ? "å›½é™…èˆªç­ ğŸŒ" : "å›½å†…èˆªç­ ğŸ ");

                sendEmail(email,subject,message);
                msgBox.setIcon(QMessageBox::Information);
                msgBox.exec();
            } else {
                QMessageBox msgBox(this);
                msgBox.setWindowTitle("Failed");
                msgBox.setText("è®¢å•æ·»åŠ å¤±è´¥: " + jsonObject["message"].toString());
                msgBox.setIcon(QMessageBox::Critical);
                msgBox.exec();
            }
        } else {
            qDebug() << "Network Error:" << reply->errorString();
            QMessageBox msgBox(this);
            msgBox.setWindowTitle("Failed");
            msgBox.setText("è®¢å•æ·»åŠ ç³»ç»Ÿè¿æ¥å¤±è´¥ !");
            msgBox.setIcon(QMessageBox::Critical);
            msgBox.exec();
        }
        reply->deleteLater();
    });
}



void pay_window::on_pushButton_4_clicked()
{
    int now_index=ui->stackedWidget->currentIndex();
    if(now_index != 0)
    {
        ui->stackedWidget->setCurrentIndex(now_index-1);
        ui->stackedWidget_2->setCurrentIndex(now_index-1);
    }
    else
    {
        InterfaceManager::instance()->switchToPage("lxt_mainInterface");
    }
}
void pay_window::sendEmail(const QString& to, const QString& subject, const QString& body) {
    QNetworkAccessManager* networkManager = new QNetworkAccessManager();
    QNetworkRequest request(QUrl("http://127.0.0.1:8080/api/send"));
    request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");

    // æ„å»ºJSONæ•°æ®
    QJsonObject json;
    json["to"] = to;
    json["subject"] = subject;
    json["body"] = body;

    QJsonDocument doc(json);
    qDebug() << "Email Request Body:" << doc.toJson();

    // å‘é€è¯·æ±‚
    QNetworkReply* reply = networkManager->post(request, doc.toJson());

    connect(reply, &QNetworkReply::finished, [reply, networkManager]() {
        QByteArray responseData = reply->readAll();
        qDebug() << "Email Response:" << responseData;

        if (reply->error() == QNetworkReply::NoError) {
            QJsonDocument jsonResponse = QJsonDocument::fromJson(responseData);
            QJsonObject jsonObject = jsonResponse.object();

            if (jsonObject["code"].toInt() == 200) {
                QMessageBox::information(nullptr, "æˆåŠŸ", "é‚®ä»¶å‘é€æˆåŠŸ");
            } else {
                QMessageBox::critical(nullptr, "å¤±è´¥",
                                      "é‚®ä»¶å‘é€å¤±è´¥: " + jsonObject["message"].toString());
            }
        } else {
            QMessageBox::critical(nullptr, "é”™è¯¯",
                                  "ç½‘ç»œé”™è¯¯: " + reply->errorString());
        }

        reply->deleteLater();
        networkManager->deleteLater();
    });
}
